ENH: Override fit method of Laplace distribution with Maximum Likelihood Estimates

swallan opened the pull request:
This pull request replaces default optimizer with maximum likelihood estimates for scale and loc of the laplace distribution. It falls back on optimizer if a parameter is set and does not match the MLE.

<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

Adds MLE equations to Laplace distribution, part of the #11782 issue. @mdhaber 

#### What does this implement/fix?
<!--Please explain your changes.-->

Implements MLE equations for laplace distribution according to equations found in the text _[Evans, Hastings, and Peacock (2000), Statistical Distributions, 3rd. Ed., John Wiley and Sons.pdf](http://personalpages.to.infn.it/~zaninett/pdf/statistical-distributions.pdf)_, page 124.

#### Additional information
<!--Any additional information you think is important.-->

Currently, when fit is called it first calculates the the MLE for scale and loc. If floc or fscale were provided, but do not match the MLE equations, then it call's the super fit method with the same fixed parameter, while raising a RunTimeWarning. I am wondering if the logic of this is efficient enough, and if it is worth checking the closeness of a supplied parameter and MLE, what the tolerance should be?

I manually tested the MLE results of laplace.rvs data with the log likelihood func from optimizer and they looked to be as good / better than those that the optimizer was able to find. It is much faster. Is there anywhere that tests should be added to ensure this behavior?

swallan commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+            scale += np.abs(data[i] - loc)
+        scale = scale/n
+
+        if ((floc is None or np.isclose(floc, loc)) and
+                (fscale is None or np.isclose(fscale, scale))):
```
The case where both floc and fscale are set and match MLE would technically pass this if statement, but are already filtered out earlier in the method.

mdhaber replied:
OK.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+            # Raise warning to indicate optimizer usage for loc or scale
+            if floc is not None:
+                # use optimizer to determine scale for custom floc
+                msg = ("Supplied floc did not match the maxmimum "
```
```suggestion
                msg = ("Supplied `floc` did not match the maximum "
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+        scale = 0
+        n = len(data)
+        for i in range(n):
+            scale += np.abs(data[i] - loc)
```
```suggestion
            scale = np.sum(np.abs(data - loc))
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+
+        scale = 0
+        n = len(data)
+        for i in range(n):
```
```suggestion
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+        Notes
+        -----
+        An error is raised if floc and fscale are given. A warning is raised
+        if a provided parameter does not match MLE equations to indicate that
```
```suggestion
        if either `floc` or `fscale` is provided but does not match the exact MLE value;
```
in this case `fit` falls back to the generic implementation.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+        data = np.asarray(data)
+
+        if not np.isfinite(data).all():
+            raise RuntimeError("The data contains non-finite values.")
```
Let's put all this boilerplate stuff in a function so we can reuse it when you override other `fit` methods.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+
+        # MLE for the laplace distribution
+        # --------------------------------
+        # scale is given by the equation
```
Let's remove these and just cite the source.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+        # Source: Statistical Distributions, 3rd Edition. Evans, Hastings,
+        #        and Peacock (2000),(pg. 124)
+        # If either floc or fscale are provided but do not match MLE
+        # equations, the optimizer will be used to determine the other value.
```
"the optimizer"
No need to assume that the reader knows what this is. Just explain that it falls back to the generic implementation in `rv_continuous`.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,6 +4195,101 @@ def _stats(self):
+            if floc is not None:
+                # use optimizer to determine scale for custom floc
+                msg = ("Supplied floc did not match the maxmimum "
+                       "likelihood estimate equation, using optimizer "
```
```suggestion
                       "likelihood estimate; using numerical optimization to determine `scale`."
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4219,49 +4219,25 @@ def fit(self, data, *args, **kwds):
-        #    \frac{1}{n}\sum_{i=0}^{n}{|x_i - loc|}
         # loc is given by
-        #    data.median()
-        # Source: Statistical Distributions, 3rd Edition. Evans, Hastings,
```
I think we said it was good to keep the source, right?

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,10 +4195,99 @@ def _stats(self):
+            if floc is not None:
+                # use optimizer to determine scale for custom floc
+                msg = ("Supplied `floc` did not match the analytical maximum "
+                       "likelihood estimate equation; using numerical "
```
```suggestion
                       "likelihood estimate; using numerical "
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,10 +4195,99 @@ def _stats(self):
+            if fscale is not None:
+                # use optimizer to determine loc for custom fscale
+                msg = ("Supplied `fscale` did not match the analytical "
+                       "maximum likelihood estimate equation;"
```
```suggestion
                       "maximum likelihood estimate;"
```

swallan replied:
^These edits are in the commit with the stacklevel change that I haven't pushed until I know that it's the right stacklevel use

ilayn commented:
The warnings need the stacklevel arguments such that the generated warnings originate from the user calling code and not within scipy code.

swallan commented:
Yes, I can confirm that @mdhaber's comments are correct. 

For example,`func` returns the same value for MLE and the optimizer values, in debug mode, having stepped into `distn_infrastructure.py`: 
<img width="520" alt="Screen Shot 2020-05-10 at 1 21 29 AM" src="https://user-images.githubusercontent.com/44255917/81494291-9b65ad80-925c-11ea-826a-a8013640d8ee.png">

swallan commented:
> The warnings need the `stacklevel `arguments such that the generated warnings originate from the user calling code and not within scipy code.

Thanks @ilayn , I see varying uses of stacklevel across scipy, and I had success with `stacklevel=2`, is this the result we are looking for?
<img width="606" alt="Screen Shot 2020-05-10 at 1 52 36 AM" src="https://user-images.githubusercontent.com/44255917/81494888-0add9c00-9261-11ea-9185-63ff5bbc7871.png">

mdhaber commented:
Thanks for the reminder @ilayn. Yes, that looks right, @swallan. We try to make it look like the error/warming is raised by the function called by the user rather than a function called by that function.

WarrenWeckesser commented:
@swallan, thanks for this PR--using explicit formulas for the parameters when they are available is a nice enhancement.

You wrote

>  If floc or fscale were provided, but do not match the MLE equations, then it call's the super fit method with the same fixed parameter, while raising a RunTimeWarning. I am wondering if the logic of this is efficient enough, and if it is worth checking the closeness of a supplied parameter and MLE, what the tolerance should be?

This is not how `floc` and `fscale` are to be handled.  If one of them is given, the user is saying "I *know* my sample is from a distribution with the given parameter; use MLE to estimate the other parameter."

Fortunately, in this case the restricted distributions are easy to handle.  For consistency with old established behavior, if *both* `floc` and `fscale` are given, we raise an error.  If just `fscale` is given, then the location is estimated by the median of `data`, and we're done.  (This works because the MLE for the location is independent of the scale.)  If just `floc` is given, then the scale is the mean absolute deviation of the data from the given `floc`.  In other words, whether `loc` is estimated as the median or given by the user, the scale (if not provided by the user) is the mean absolute deviation of the data from `loc`.

I have implement the MLE for Laplace in my [`mpsci`](https://github.com/WarrenWeckesser/mpsci) library. The [`mle` function](https://github.com/WarrenWeckesser/mpsci/blob/0926e685ba5eb9183b6c49173def26139b990017/mpsci/distributions/laplace.py#L140) is in the file [`'laplace.py'`](https://github.com/WarrenWeckesser/mpsci/blob/master/mpsci/distributions/laplace.py).  The code in `mpsci` doesn't raise an error if both the parameters are fixed--it just returns the given parameters without doing anything--so the logic here might be a little different than what we use in the scipy code, but it shows one way to handle fixed parameter values.  (The names are different: setting `floc` or `fscale` in the scipy code corresponds to setting `mu` or `b`, respectively, in the mpsci code.)

Here's a simple test case:  suppose `data = np.array([1.0, 1.0, 3.0, 5.0, 8.0, 14.0])`.   The unrestricted MLE for this data is `loc=4` and `scale=11/3`.  With `floc=6`, however, the estimate of the scale should be 4.

mdhaber commented:
> This is not how floc and fscale are to be handled.

I think that's my fault. I didn't stop to think about whether the analytical formulas could still be used.

> If one of them is given, the user is saying "I know my sample is from a distribution with the given parameter; use MLE to estimate the other parameter."

Using the `super` method to do the job is not inconsistent with that, though...

WarrenWeckesser commented:
There is no need to fall back on the generic implementation here.  All the cases can be handled explicitly.

WarrenWeckesser commented:
In pseudocode:
```
if both floc and fscale given:
    error

if floc given:
    loc = floc
else:
    loc = median(data)

if fscale given:
    scale = fscale
else:
    scale = mean absolute deviation of data from loc

return loc, scale
```

swallan commented:
> There is no need to fall back on the generic implementation here. All the cases can be handled explicitly.

@WarrenWeckesser Thank you for clarifying!

mdhaber commented:
@swallan During our meeting we discussed a possibility described below. Consider the following:
1. While holding the location fixed at a given value, you fit the scale using numerical optimization (default implementation).
2. Given the value of the location, fit the scale according to the analytical expression.
Can the value of the objective function in 1 be lower than the objective function in 2.

I think you mentioned that you had a numerical example? Can you provide it?

---

I would have assumed this was possible because the analytical expression for the scale was derived when there were two decision variables. For a general optimization problem, I would not assume that:

(a) a formula for the optimal value of a parameter (e.g. the scale) that is derived when two parameters are free (e.g. the location and scale) would be the same as
(b) a formula for the optimal value of a parameter (e.g. the scale) that is derived when only that parameter is free and a second parameter (e.g. the location) is an arbitrary value.

That's why I originally suggested that you fall back on the default numerical implementation, fixing one parameter, if one parameter is given. I didn't think the formula for the second parameter would be valid.

However, it sounds like @WarrenWeckesser has looked into the math and found that _in this case_, (a) and (b) above are in fact the same. The problems happen to decouple.

swallan commented:
@mdhaber I must have been mistaken, I thought that I had seen this happen but wasn't able to replicate it. It appears that the optimal value is always given by the MLE formula, fixed parameters or not. 

I tested this by individually fixing both `floc` and `fscale` with mle and the optimizer. The results were the same. 
![Screen Shot 2020-05-11 at 4 31 44 PM](https://user-images.githubusercontent.com/44255917/81622076-eab6f580-93a4-11ea-9e8b-0b4d9842247c.png)

mdhaber commented:
OK! Good to know.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -1374,6 +1374,56 @@ def test_stats2(self):
+        loc_opt, scale_opt = super(type(stats.laplace),
+                                   stats.laplace).fit(data)
+        assert_allclose(loc_opt, loc, atol=.01)
+        assert_allclose(scale_opt, scale, atol=.01)
```
Let's replace these assertions with the assertion that the analytical value should be <= the numerically optimized value.

swallan replied:
> assertion that the analytical value should be <= the numerically optimized value.

May I confirm that this is to mean that the objective function result of the analytical value is <= to the objective function result of the numerically optimized value? I think this is what you indicated in the rest of your comment.

mdhaber replied:
Yes, my fault.

swallan replied:
No problem!

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -1374,6 +1374,56 @@ def test_stats2(self):
+    def test_fit_MLE_comp_optimzer(self, rvs_loc, rvs_scale):
+        data = stats.laplace.rvs(size=1000, loc=rvs_loc, scale=rvs_scale)
+
+        # that MLE estimates are almost equal to optimizer
```
Be more precise:
```suggestion
        # test that the analytical MLEs are close to the numerically optimized MLEs
```

mdhaber replied:
Of course this will need to change if you replace the assertions. The point is that using "MLE estimates" to refer to the _analytical estimates_ and "optimizer" to refer to the _numerically optimized estimates_ is ambiguous.

swallan replied:
I will be more specific with the language.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -1374,6 +1374,56 @@ def test_stats2(self):
+        assert_allclose(loc_opt, loc, atol=.01)
+        assert_allclose(scale_opt, scale, atol=.01)
+
+    def test_fixed_parameter(self):
```
Is "test_fixed_parameter" the most appropriate name? You've already tested with fixed parameters in `test_fit`, so what is the purpose of this test?

swallan replied:
It could probably be better titled to indicate that it is a non-random data sample. Is `test_fit_simple_non_random_data` more appropriate? I included this test because it was simple to see the effects of fixing a parameter with simpler data.

mdhaber replied:
Sounds good to me.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -1374,6 +1374,56 @@ def test_stats2(self):
 
+    @pytest.mark.parametrize("rvs_loc", [-5, 0, 1, 2])
+    @pytest.mark.parametrize("rvs_scale", [1, 2, 3, 10])
+    def test_fit(self, rvs_loc, rvs_scale):
```
Fix the RNG seed.

swallan replied:
I see 
``` 
def setup_method(self):
        np.random.seed(1234)
``` 
inside the `TestDLaplace` class, should I set the seed inside each method?

mdhaber replied:
Yes - well, at least if there is something random going on, like `rvs`. Also, check at the top of the document. If it says `from numpy.random import seed` or something, just use `seed(1234)`. (I remembar Ilayn asking us for that on the LAPACK work.)

mdhaber replied:
I'm sorry @swallan. Somehow I didn't read the bit about `setup_method`... it's been a really busy week. It looks like `setup_method` runs before the method runs, you're absolutely right, the seed does not need to be set every time. Sorry!

swallan replied:
No problem @mdhaber, fixed!

WarrenWeckesser commented:
Matt and Sam,

Sorry for some of my very terse comments earlier. Now I'll probably err in the other direction and say too much...

In my previous comments, I was focused on Laplace specifically. In general, Matt's points are correct:

* For other distributions, we won't necessarily have explicit formulas for all the possible combinations of some parameters being fixed and some not.  When we don't have a formula, we fall back to the numerical optimization.  Because of the way SciPy continuous distributions are parametrized (i.e. they *all* have location and scale parameters), I suspect a typical case will be that we don't have a formula for the case where no parameters are fixed, but we have one for, say, the location being fixed.  That's the case for the gamma distribution, for example: if the location is not fixed, `gamma_gen.fit` falls back on the generic optimization. 
* In general (i.e. for other distributions), if we have an explicit formula for the case where no parameters are fixed, we can't necessarily use those same formula when one or more parameters are fixed.  Whether or not that works depends on the specific properties of the log-likelihood function of that distribution. In the specific case of Laplace, it works.

-----

Some more comments on the MLE for Laplace (for conciseness, I'll use m for the location parameter and b for the scale parameter):

The log-likelihood function for the Laplace distribution for the parameters m and b, and vector of values x with length n, is

    ll(m, b; x) = -n*log(2*b) - (1/b)*sum(abs(x[i] - m))

Usually to find the extreme values of such function, one would take the partial derivatives with respect to m and b, set them equal to 0, and solve for m and b to find the candidate extreme values.  The partial with respect to b is

    -n/b + (1/b**2)*sum(abs(x[i] - m))

Set that equal to 0; regardless of whether m is fixed or free, we can then solve for b to obtain:

    b = sum(abs(x[i] - m))/n

Technically, we can't differentiate ll with respect to m because the absolute value function is not differentiable at 0.  In this case, it is worthwhile to look at the behavior of the expression sum(abs(x[i] - m)) considered as a function of m.  The following plots show two examples,
one with x = [-4, -3.5, -0.5, 1.2, 3.5], so len(x) = 5 and one with x = [-4, -3.5, -0.5, 1.2, 3.5, 4.1], so len(x) = 6.

![Figure_1](https://user-images.githubusercontent.com/321463/81717522-a846f480-9448-11ea-92c0-3d84f28f5d00.png)


When len(x) is odd, there is a unique minimum at the median of x.  When len(x) is even, graph has a flat segment--any value of m here minimizes sum(abs(x[i] - m)).  The conventional definition of the median is to pick the midpoint of this interval, and that corresponds to the midpoint of the flat region of the graph.  So median(x) is *a* minimizer of sum(abs(x[i] - m)); when len(x) is odd, it is the unique minimizer.

Now look at ll(m, b; x); it has the form F(b) - G(b)*H(m), where G(b) > 0 and H(m) >= 0 for all m.  If (b, m) maximizes ll, them m must minimize H(m).  So we can choose m = median(x) as a MLE for Laplace.  This argument applies whether we consider b fixed or free.

The conclusion of all this is summarized in the pseudocode in my previous comment.

mdhaber commented:
@WarrenWeckesser thanks. I think @swallan has responded to your comments.

swallan commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -1374,6 +1374,80 @@ def test_stats2(self):
         assert_equal((m, s), (0., 0.))
         assert_allclose((v, k), (4., 3.25))
 
+    @pytest.mark.parametrize("rvs_loc", [-5, 0, 1, 2])
```
Thanks for catching that! I didn't know it was a different class and not just an interesting naming choice.

swallan commented:
@WarrenWeckesser Thanks for the specific feedback, and I think that you are right about using `assert_allclose(...., atol=1e-15, rtol=1e-15)` instead of `assert_equal`.  Similarly, a random failure popped up with a `<=` that I had used, so I modified that to use np.allclose with those tolerances.

I also added the note to the docstring, and addressed the inline changes you suggested.

tylerjereddy commented:
@WarrenWeckesser Are you happy here now?

WarrenWeckesser commented:
@tylerjereddy, I just got back to this a few minutes ago.  I give an update ASAP.

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,10 +4195,82 @@ def _stats(self):
 
+    @replace_notes_in_docstring(rv_continuous, notes="""\
+        This function uses explicit formulas for the maximum likelihood
+        estimation of the laplace distribution parameters, so the keyword
```
```suggestion
        estimation of the Laplace distribution parameters, so the keyword
```

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,10 +4195,82 @@ def _stats(self):
+        estimation of the laplace distribution parameters, so the keyword
+        arguments `loc`, `scale`, and `optimizer` are ignored.\n\n""")
+    def fit(self, data, *args, **kwds):
+        """
+        Maximum likelihood estimate for the location and scale parameters.
+
+        Exact formulas are used to calculate the location and scale, unless
+        they are provided in `floc` or `fscale`.
+
+        Parameters
+        ----------
+        data : array_like
+            Data to use in calculating the maximum likelihood estimate.
+        floc : float, optional
+            Hold the location parameter fixed to the specified value.
+        fscale : float, optional
+            Hold the scale parameter fixed to the specified value.
+
+        Returns
+        -------
+        loc, scale : float
+            Maximum likelihood estimates for the location and scale.
+
+        Notes
+        -----
+        An error is raised if both `floc` and `fscale` are fixed. Explicit
+        formulas are used, so the keyword arguments `loc`, `scale`, and
+        `optimizer` are ignored.
+        """
```
```suggestion
```

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,10 +4195,82 @@ def _stats(self):
+        floc = kwds.pop('floc', None)
+        fscale = kwds.pop('fscale', None)
+
+        check_fit_input_parameters(data, args,
+                                   kwds, fixed_param=(floc, fscale))
```
```suggestion
        _check_fit_input_parameters(data, args,
                                    kwds, fixed_param=(floc, fscale))
```

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4195,10 +4195,82 @@ def _stats(self):
 laplace = laplace_gen(name='laplace')
 
 
+def check_fit_input_parameters(data, args, kwds, fixed_param):
```
```suggestion
def _check_fit_input_parameters(data, args, kwds, fixed_param):
```

WarrenWeckesser commented:
When the tests finish, I'll squash and merge.

WarrenWeckesser commented:
Merged.  Thanks @swallan, and thanks @mdhaber for reviewing.

mdhaber commented:
I'm taking a second pass through gh-11782, so I thought I should report the results of a test of this fit override. Methods will be described in gh-11782.

For all combinations of fixed/free, this PR is faster. Typically, something like:
![image](https://user-images.githubusercontent.com/6570539/192123587-6d3ab709-3e08-43dd-99a6-8ef20954a6bb.png)

It always improves (or maintains when numerical optimization is very good) the NLLF of the fit:
![image](https://user-images.githubusercontent.com/6570539/192123609-5e1f9359-2cba-4867-b7c7-9b2e0cb54f76.png)

I don't think we need to revisit this; it's looking very good.
